File: backend/internal/ndrclient/assets.go
package ndrclient

import (
	"context"
	"fmt"
	"net/http"
	"time"
)

// Asset represents a file asset stored in NDR.
type Asset struct {
	ID          int64      `json:"id"`
	Filename    string     `json:"filename"`
	ContentType *string    `json:"content_type,omitempty"`
	SizeBytes   int64      `json:"size_bytes"`
	Status      string     `json:"status"`
	Bucket      string     `json:"bucket"`
	ObjectKey   string     `json:"object_key"`
	ETag        *string    `json:"etag,omitempty"`
	CreatedBy   string     `json:"created_by"`
	UpdatedBy   string     `json:"updated_by"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	DeletedAt   *time.Time `json:"deleted_at,omitempty"`
}

// AssetInitRequest is the request body for initializing a multipart upload.
type AssetInitRequest struct {
	Filename    string `json:"filename"`
	ContentType string `json:"content_type,omitempty"`
	SizeBytes   int64  `json:"size_bytes"`
}

// AssetInitResponse is the response from initializing a multipart upload.
type AssetInitResponse struct {
	Asset         Asset  `json:"asset"`
	UploadID      string `json:"upload_id"`
	PartSizeBytes int    `json:"part_size_bytes"`
	ExpiresIn     int    `json:"expires_in"`
}

// AssetPartURLsRequest is the request body for getting presigned part URLs.
type AssetPartURLsRequest struct {
	PartNumbers []int `json:"part_numbers"`
}

// AssetPartURL represents a presigned URL for uploading a part.
type AssetPartURL struct {
	PartNumber int    `json:"part_number"`
	URL        string `json:"url"`
}

// AssetPartURLsResponse is the response from getting presigned part URLs.
type AssetPartURLsResponse struct {
	UploadID  string         `json:"upload_id"`
	URLs      []AssetPartURL `json:"urls"`
	ExpiresIn int            `json:"expires_in"`
}

// AssetCompletedPart represents a completed upload part.
type AssetCompletedPart struct {
	PartNumber int    `json:"part_number"`
	ETag       string `json:"etag"`
}

// AssetCompleteRequest is the request body for completing a multipart upload.
type AssetCompleteRequest struct {
	Parts []AssetCompletedPart `json:"parts"`
}

// AssetDownloadURLResponse is the response from getting a download URL.
type AssetDownloadURLResponse struct {
	URL       string `json:"url"`
	ExpiresIn int    `json:"expires_in"`
}

// InitMultipartUpload initializes a multipart upload session.
func (c *httpClient) InitMultipartUpload(ctx context.Context, meta RequestMeta, req AssetInitRequest) (AssetInitResponse, error) {
	httpReq, err := c.newRequest(ctx, http.MethodPost, "/api/v1/assets/multipart/init", meta, req)
	if err != nil {
		return AssetInitResponse{}, err
	}
	var resp AssetInitResponse
	_, err = c.do(httpReq, &resp)
	return resp, err
}

// GetAssetPartURLs gets presigned URLs for uploading parts.
func (c *httpClient) GetAssetPartURLs(ctx context.Context, meta RequestMeta, assetID int64, partNumbers []int) (AssetPartURLsResponse, error) {
	endpoint := fmt.Sprintf("/api/v1/assets/%d/multipart/part-urls", assetID)
	req := AssetPartURLsRequest{PartNumbers: partNumbers}
	httpReq, err := c.newRequest(ctx, http.MethodPost, endpoint, meta, req)
	if err != nil {
		return AssetPartURLsResponse{}, err
	}
	var resp AssetPartURLsResponse
	_, err = c.do(httpReq, &resp)
	return resp, err
}

// CompleteMultipartUpload completes a multipart upload.
func (c *httpClient) CompleteMultipartUpload(ctx context.Context, meta RequestMeta, assetID int64, parts []AssetCompletedPart) (Asset, error) {
	endpoint := fmt.Sprintf("/api/v1/assets/%d/multipart/complete", assetID)
	req := AssetCompleteRequest{Parts: parts}
	httpReq, err := c.newRequest(ctx, http.MethodPost, endpoint, meta, req)
	if err != nil {
		return Asset{}, err
	}
	var resp Asset
	_, err = c.do(httpReq, &resp)
	return resp, err
}

// AbortMultipartUpload aborts a multipart upload.
func (c *httpClient) AbortMultipartUpload(ctx context.Context, meta RequestMeta, assetID int64) error {
	endpoint := fmt.Sprintf("/api/v1/assets/%d/multipart/abort", assetID)
	httpReq, err := c.newRequest(ctx, http.MethodPost, endpoint, meta, nil)
	if err != nil {
		return err
	}
	_, err = c.do(httpReq, nil)
	return err
}

// GetAsset gets asset metadata by ID.
func (c *httpClient) GetAsset(ctx context.Context, meta RequestMeta, assetID int64) (Asset, error) {
	endpoint := fmt.Sprintf("/api/v1/assets/%d", assetID)
	httpReq, err := c.newRequest(ctx, http.MethodGet, endpoint, meta, nil)
	if err != nil {
		return Asset{}, err
	}
	var resp Asset
	_, err = c.do(httpReq, &resp)
	return resp, err
}

// GetAssetDownloadURL gets a presigned download URL for an asset.
func (c *httpClient) GetAssetDownloadURL(ctx context.Context, meta RequestMeta, assetID int64) (AssetDownloadURLResponse, error) {
	endpoint := fmt.Sprintf("/api/v1/assets/%d/download-url", assetID)
	httpReq, err := c.newRequest(ctx, http.MethodGet, endpoint, meta, nil)
	if err != nil {
		return AssetDownloadURLResponse{}, err
	}
	var resp AssetDownloadURLResponse
	_, err = c.do(httpReq, &resp)
	return resp, err
}

// DeleteAsset soft-deletes an asset.
func (c *httpClient) DeleteAsset(ctx context.Context, meta RequestMeta, assetID int64) error {
	endpoint := fmt.Sprintf("/api/v1/assets/%d", assetID)
	httpReq, err := c.newRequest(ctx, http.MethodDelete, endpoint, meta, nil)
	if err != nil {
		return err
	}
	_, err = c.do(httpReq, nil)
	return err
}

File: frontend/src/api/assets.ts
import { http } from "./http";

/**
 * Asset 资产类型定义
 */
export interface Asset {
  id: number;
  filename: string;
  content_type: string | null;
  size_bytes: number;
  status: string;
  bucket: string;
  object_key: string;
  etag: string | null;
  created_by: string;
  updated_by: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

/**
 * 初始化上传请求参数
 */
export interface AssetInitRequest {
  filename: string;
  content_type?: string;
  size_bytes: number;
}

/**
 * 初始化上传响应
 */
export interface AssetInitResponse {
  asset: Asset;
  upload_id: string;
  part_size_bytes: number;
  expires_in: number;
}

/**
 * 分片 URL
 */
export interface AssetPartURL {
  part_number: number;
  url: string;
}

/**
 * 分片 URL 响应
 */
export interface AssetPartURLsResponse {
  upload_id: string;
  urls: AssetPartURL[];
  expires_in: number;
}

/**
 * 已完成的分片信息
 */
export interface CompletedPart {
  part_number: number;
  etag: string;
}

/**
 * 下载 URL 响应
 */
export interface AssetDownloadURLResponse {
  url: string;
  expires_in: number;
}

/**
 * 初始化分片上传
 */
export async function initMultipartUpload(
  params: AssetInitRequest
): Promise<AssetInitResponse> {
  return http<AssetInitResponse>("/api/v1/assets/multipart/init", {
    method: "POST",
    body: JSON.stringify(params),
  });
}

/**
 * 获取分片预签名 URL
 */
export async function getPartURLs(
  assetId: number,
  partNumbers: number[]
): Promise<AssetPartURLsResponse> {
  return http<AssetPartURLsResponse>(
    `/api/v1/assets/${assetId}/multipart/part-urls`,
    {
      method: "POST",
      body: JSON.stringify({ part_numbers: partNumbers }),
    }
  );
}

/**
 * 完成分片上传
 */
export async function completeMultipartUpload(
  assetId: number,
  parts: CompletedPart[]
): Promise<Asset> {
  return http<Asset>(`/api/v1/assets/${assetId}/multipart/complete`, {
    method: "POST",
    body: JSON.stringify({ parts }),
  });
}

/**
 * 中止分片上传
 */
export async function abortMultipartUpload(assetId: number): Promise<void> {
  return http<void>(`/api/v1/assets/${assetId}/multipart/abort`, {
    method: "POST",
  });
}

/**
 * 获取资产信息
 */
export async function getAsset(assetId: number): Promise<Asset> {
  return http<Asset>(`/api/v1/assets/${assetId}`);
}

/**
 * 获取下载 URL
 */
export async function getDownloadURL(
  assetId: number
): Promise<AssetDownloadURLResponse> {
  return http<AssetDownloadURLResponse>(
    `/api/v1/assets/${assetId}/download-url`
  );
}

/**
 * 删除资产
 */
export async function deleteAsset(assetId: number): Promise<void> {
  return http<void>(`/api/v1/assets/${assetId}`, {
    method: "DELETE",
  });
}

/**
 * 上传单个分片到 S3 (直接使用 presigned URL，不通过 http())
 * @param url presigned URL
 * @param blob 文件数据
 * @param onProgress 进度回调
 * @returns ETag
 */
export async function uploadPart(
  url: string,
  blob: Blob,
  onProgress?: (loaded: number) => void
): Promise<string> {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();

    xhr.upload.addEventListener("progress", (event) => {
      if (event.lengthComputable && onProgress) {
        onProgress(event.loaded);
      }
    });

    xhr.addEventListener("load", () => {
      if (xhr.status >= 200 && xhr.status < 300) {
        // S3 返回的 ETag 在响应头中
        const etag = xhr.getResponseHeader("ETag");
        if (etag) {
          // 移除引号
          resolve(etag.replace(/"/g, ""));
        } else {
          reject(new Error("Missing ETag in response"));
        }
      } else {
        reject(new Error(`Upload failed with status ${xhr.status}`));
      }
    });

    xhr.addEventListener("error", () => {
      reject(new Error("Upload failed"));
    });

    xhr.addEventListener("abort", () => {
      reject(new Error("Upload aborted"));
    });

    xhr.open("PUT", url);
    xhr.send(blob);
  });
}
